semver/validity.ts

```diff
@@ -0,0 +1,162 @@
+      isValidString(v)
+    ) &&
+    prerelease.filter((v) => typeof v === "number").every((v) =>
+      isValidNumber(v)
+    ) &&
+    build.every((v) => typeof v === "string" && isValidString(v));
+  return result;
+}
+
+/**
+ * Does a deep check on the value to see if it is a valid SemVerComparator object.
+ *
+ * Objects with extra fields are still considered valid if they have at
+ * least the correct fields.
+ *
+ * Adds a type assertion if true.
+ * @param value The value to check if its a SemVerComparator
+ * @returns True if the object is a SemVerComparator otherwise false
+ */
+export function isSemVerComparator(value: unknown): value is SemVerComparator {
+  if (value == null) return false;
+  if (value === NONE) return true;
+  if (value === ALL) return true;
+  if (Array.isArray(value)) return false;
+  if (typeof value !== "object") return false;
+  const { operator, semver, min, max } = value as SemVerComparator;
+  return (
+    isValidOperator(operator) &&
+    isSemVer(semver) &&
+    isSemVer(min) &&
+    isSemVer(max)
+  );
+}
+
+/**
+ * Does a deep check on the object to determine if its a valid range.
+ *
+ * Objects with extra fields are still considered valid if they have at
+ * least the correct fields.
+ *
+ * Adds a type assertion if true.
+ * @param value The value to check if its a valid SemVerRange
+ * @returns True if its a valid SemVerRange otherwise false.
+ */
+export function isSemVerRange(value: unknown): value is SemVerRange {
+  if (value == null) return false;
+  if (Array.isArray(value)) return false;
+  if (typeof value !== "object") return false;
+  const { ranges } = value as SemVerRange;
+  return (
+    Array.isArray(ranges),
+      ranges.every((r) =>
+        Array.isArray(r) && r.every((c) => isSemVerComparator(c))
+      )
+  );
+}
+
+/**
+ * Checks to see if the value is a valid Operator string.
+ *
+ * Adds a type assertion if true.
+ * @param value The value to check
+ * @returns True if the value is a valid Operator string otherwise false.
+ */
+export function isValidOperator(value: unknown): value is Operator {
+  if (typeof value !== "string") return false;
+  switch (value) {
+    case "":
+    case "=":
+    case "==":
+    case "===":
+    case "!==":
+    case "!=":
+    case ">":
+    case ">=":
+    case "<":
+    case "<=":
+      return true;
+    default:
+      return false;
+  }
+}
+
+/**
+ * Returns true if the value is a valid SemVer number.
+ *
+ * Must be a number. Must not be NaN. Can be positive or negative infinity.
+ * Can be between 0 and MAX_SAFE_INTEGER.
+ * @param value The value to check
+ * @returns True if its a valid semver number
+ */
+export function isValidNumber(value: unknown): value is number {
+  return (
+    typeof value === "number" &&
+    !isNaN(value) && (
+      value === Number.POSITIVE_INFINITY ||
+      value === Number.NEGATIVE_INFINITY ||
+      (
+        value <= Number.MAX_SAFE_INTEGER &&
+        value >= 0
+      )
+    )
```

Comment:
Nit:

After eliminating `NaN` as a possibility (see [`Number.isNaN()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isNaN) vs [global `isNaN()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/isNaN)), testing for `±∞` can be collapsed into a single functional check by negating [`Number.isFinite()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isFinite)

```suggestion
    !Number.isNaN(value) && (
      !Number.isFinite(value) ||
      (0 <= value && value <= Number.MAX_SAFE_INTEGER)
    )
```